Feat: "RunningGuideView 화면 구성" 및 일부 코드 수정

wh5929 opened the pull request:
기존에 작업중이던 branch가 github에서 표시가 안되고 PR 화면에서 해당 브랜치를 선택할 수 없는 증상이 있어
develop에서 create branch 후 코드를 추가했습니다.
따라서 코드가 추가/삭제/수정된 표시를 이번 PR에서 확인하시기 어려우실듯 합니다.

신속한 작업을 위하여 3일 내에 PR 검토되지 않을 경우 merge 진행하겠습니다.


### 기능 추가
1. RunningGuideView 화면 구성
2. Custom Scroll 구현
3. 바로 시작 뷰(JustStartView)의 runGuideExperienceTabItem 탭 이벤트 발생시 해당 항목을 fullScreenCover를 이용해 상세화면 표시
4. Show+Extension.swift Util 추가 -> Bool 프로퍼티를 이용해 View를 표시/숨기도록 하는 ViewModifier와 extension


### 코드 수정
1. RunGuideItem 프로퍼티 항목 추가 및 이름 변경
2. RunGuideViewModel 에서 추가된 프로퍼티 항목 기입 및 내용 추가 / 불필요한 코드 삭제
3. 필요없거나 저작권 있는 사진 삭제 / Asset 사진 이름 변경
4. 현재 M1 Pro의 시뮬레이터에서 iPhone 14로 GatheRunner 프로젝트를 실행했을때  바로 시작 뷰(JustStartView)의 Map 뷰 인스턴스로 인해  profiler 상에서 CPU 사용량 30%가 되는것을 확인
-> 이것 때문인지 바로 시작 뷰(JustStartView)에서 runningGuideTabItem 항목을 탭해서 fullScreenCover가 표시되었을때 스크롤이 버벅이는 증상이 있음
-> 바로 시작 뷰(JustStartView)의 Map을 Show+Extension을 이용해 fullScreenCover가 표시되었을때 .hidden() 보이지 않게 설정하고 fullScreenCover가 dismiss() 해제 되었을때 다시 보이도록 설정

wh5929 committed:
### 기능 추가
1. RunningGuideView 화면 구성
2. Custom Scroll 구현
3. 바로 시작 뷰(JustStartView)의 runGuideExperienceTabItem 탭 이벤트 발생시 해당 항목을 fullScreenCover를 이용해 상세화면 표시
4. Util에 Show+Extension.swift 파일 추가 -> Bool 프로퍼티를 이용해 View를 표시/숨기도록 하는 ViewModifier와 extension

### 내용 수정
1. RunGuideItem 프로퍼티 항목 추가 및 이름 변경
2. RunGuideViewModel 에서 추가된 프로퍼티 항목 기입 및 내용 추가 / 불필요한 코드 삭제
3. 필요없거나 저작권 있는 사진 삭제 / Asset 사진 이름 변경

DalioKim commented:
**RunGuideExperienceTabItem**

[의견] 
1. vStackFrameWidthRatio등 리터럴 값을 상수로 선언하는 방식이 일반적이지 않습니다. 저번에 공유해드렸듯이 네임스페이스를 이용해서 선언하는게 좋아보입니다. 그리고 변수명 수정도 필요해보입니다. vStack이라고 표현하기보단 어떤 스택뷰인지 명명(ex: textLayer)하시고 네이밍 + WidthRatio(ex: textLayerWidthRatio)로 하시는게 나아보입니다. 
2. arrsRunGuideExperienceTabItem 이 변수는 왜 3개의 view에서 따로따로 선언되는지 모르겠습니다. 
3. TabItemImages, TabItemTexts은 타입으로 분류할 필요가 없어보입니다. 변수로 선언후 분리를 유지하셔도 될 것 같습니다.

DalioKim commented:
**RunningView**

[의견] 
1. 위의 의견과 같이 리터럴 값을 상수로 선언방식 수정 필요해보입니다.
2. 불린 변수명의 일반적인 컨벤션은 is + ~ 입니다. 그리고 버튼과 같이 사용되는데 변수명에 justStartButton등 모든 정보가 들어가있을 필요는 없을 것 같습니다. 
3. BtnSelector, MenuBtn 메서드나 계산속성, 연산자로 대체하셔도 될 것 같습니다. 타입으로 분리할 필요는 없어보입니다.

DalioKim commented:
**JustStartView**

[의견]
1. 위의 의견과 같이 리터럴 값을 상수로 선언방식 수정 필요해보입니다.
2. runGuideExperienceItemArrs은 뷰에서 또 프로퍼티로 선언할 필요가 없어보입니다. 그리고 Mocks 데이터는 Running/Model/extension 해당 View의 static let mockedData으로 선언하시는 편이 좋아보입니다.  
3. extension으로 구분할때드는 비슷한 속성끼리 묶으셔야합니다. runGuideTabView와 mapView가 서브뷰로써 구분되는게 더 적합해보입니다. 
4. runGuideTabView의 UI로직을 분리하는 편이 좋을 것 같습니다. 현재 선언하신 방식은 다소 가독성이 떨어져 보입니다. 

[질문]
1. selectedCardItem은 JustStartView 초기화시 파라미터로 받는데, 옵셔널타입으로 정의하시고 Nil로 선언하신 이유가 있으신가요?

DalioKim commented:
**Show+Extension**

[의견]
1. EmptyModifier와 겹칩니다. 특별한 이유가 없다면 EmptyModifier를 사용하셔도 될 것 같습니다.
2. Extension은 Helper에 분류하는게 맞아 보입니다.
3. Extension 파일의 네이밍은 대상(ex: Int+Extension)+Extension 입니다.

DalioKim commented:
**RunningGuideView**

[의견]
1. cardItem은 아이템들인데 아이템으로 명명되어 있어 혼동할 수 있을 것 같습니다. 또한 componet 들을 (cardImage, cardTitle등으로) 분리하는게 좋을 것 같습니다. 
2. 위와 같이 size 관련 상수들은 enum 네임스페이스에 정의하고, runGuideCardItemArrs은 뷰모델을 이용해서 직접 뷰에 전달하면 될 것 같습니다. 

[질문]
1. 위와 같이 selectedCardItem은 옵셔널타입 및 nil 할당 이유가 궁금합니다.

**FullScreenCoverView**

[의견]
1. 파일을 분리하는게 더 적합해보입니다. 
4. 변수명들이 모든 내용을 포함하고 있어, 가독성이 좋지 않습니다.
5. 이전 좌표값을 담는 변수들이 매번 초기화되는데 그렇다면 변수로 갖고있거나, 메서드로 계산할 필요도 없어보입니다.
6. onChange에 대한 설명을 주석으로 할 필요는 없어보입니다. 
7. FullScreenCoverView와 TextViewSetForRunGuideBottomView 네이밍은 반드시 수정될 필요가 있어보입니다. 
8. TextViewSetForRunGuideBottomView와 CloseButton은 타입으로 분리되고, 각각의 size 정보가 따로 관리될 필요가 없어보입니다. 이게 전체적으로 다른 파일에서도 이런 방식으로 구현하셨는데, 결국엔 FullScreenCoverView와 같은 컨테이너뷰에서 size가 결정될텐데 추후에 유지보수나 화면비율을 맞추는데 어려움이 있을 것 같습니다.
9. gesture, onChange등 이벤트 연산자 따로 분리하는게 좋아보입니다.

[질문]
1. CloseButton을 최상단 우측에 배치하시려고 하신것 같은데 다른 UI로직으로는 안될까요? 
2. 메서드로 분리하는 기준이 제각각 인것 같습니다. 제스쳐 연산자에 있는 로직과 setForCardItemOverlayYAxisPosition 메서드로 따로 분리된 로직은 어떤 차이가 있기에, 메서드로 분리되거나 분리하지 않은건가요?
3. getForRunGuideBottomDescriptionView은 왜 EmptyView를 리턴하는건가요?

DalioKim commented:
RunningGuideView는 확인 및 리뷰할 부분이 많아서 이 부분 이야기가 될 때까지 머지는 안하는게 좋을 것 같습니다.

wh5929 commented:
> RunGuideExperienceTabItem [의견]
> 
> 1. vStackFrameWidthRatio등 리터럴 값을 상수로 선언하는 방식이 일반적이지 않습니다. 저번에 공유해드렸듯이 네임스페이스를 이용해서 선언하는게 좋아보입니다. 그리고 변수명 수정도 필요해보입니다. vStack이라고 표현하기보단 어떤 스택뷰인지 명명하시고 네이밍 + WidthRatio로 하시는게 나아보입니다.
> 2. arrsRunGuideExperienceTabItem 이 변수는 왜 3개의 view에서 따로따로 선언되는지 모르겠습니다.
> 3. TabItemImages, TabItemTexts은 타입으로 분류할 필요가 없어보입니다. 변수로 선언후 분리를 유지하셔도 될 것 같습니다.

### 답변
1. 리터럴값을 상수로 선언하는것은 이전에 얘기가 나왔었습니다. 그러나 상수명이 적절치 않네요. V/HStack~ 으로 시작하는 어두에 객체명을 집어넣으면 좋겠네요.
<img width="898" alt="image" src="https://user-images.githubusercontent.com/10336586/195338225-6752c03d-aee7-4c47-a3e3-a711d16a0920.png">

2.3. 코드 구현부를 분리하고 나서 체크를 못했네요. extension을 사용하여 하나의 타입으로 묶었습니다.

wh5929 commented:
> **Show+Extension**
> 
> [의견]
> 
> 1. EmptyModifier와 겹칩니다. 특별한 이유가 없다면 EmptyModifier를 사용하셔도 될 것 같습니다.
> 2. Extension은 Helper에 분류하는게 맞아 보입니다.
> 3. Extension 파일의 네이밍은 대상(ex: Int+Extension)+Extension 입니다.

### 답변

1. EmptyModifier의 사용을 고려했으나 ViewObject.hidden() 설정과 EmptyView()를 반환하는 것에는 차이가 있습니다. 그래서 해당 Extension을 사용합니다.

2.3. 참고하여 수정했습니다.

DalioKim commented:
> 1. 리터럴값을 상수로 선언하는것은 이전에 얘기가 나왔었습니다. 그러나 상수명이 적절치 않네요. V/HStack~ 으로 시작하는 어두에 객체명을 집어넣으면 좋겠네요.

워딩에 다소 오해가 있는데 상수로 선언한게 잘못됬다는게 아니라 선언하는 방식이 일반적이지 않다라는 말입니다.

DalioKim commented:
> 3\. 이 부분은 계속 이야기가 나오고 있는데 동현님이 타입으로 추출하자고 해서 타입으로 만들었던 부분입니다.
>     추후 변수로 수정할 필요성이 생기면 그 때 수정하도록 하겠습니다.

제가 아마 타입으로 추출해야 할것같다고 의견드린건 featuredTabView와 같은 부분이었을겁니다. 그리고 추출하는 방식에는 타입만 있는게 아니라 프로퍼티들도 있습니다.

DalioKim commented:
> 1. EmptyModifier의 사용을 고려했으나 ViewObject.hidden() 설정과 EmptyView()를 반환하는 것에는 차이가 있습니다. 그래서 해당 Extension을 사용합니다.

hidden은 레이아웃의 공간을 차지하고 있고, EmptyView 차지하고 있지 않은 차이인데, fullScreenCover에 의해 화면이 가려지는데 어떠한 차이가 있는건가요?

wh5929 commented:
**JustStartView**

[의견]
1. 위의 의견과 같이 리터럴 값을 상수로 선언방식 수정 필요해보입니다.
2. runGuideExperienceItemArrs은 뷰에서 또 프로퍼티로 선언할 필요가 없어보입니다. 그리고 Mocks 데이터는 Running/Model/extension 해당 View의 static let mockedData으로 선언하시는 편이 좋아보입니다.  
3. extension으로 구분할때드는 비슷한 속성끼리 묶으셔야합니다. runGuideTabView와 mapView가 서브뷰로써 구분되는게 더 적합해보입니다. 
4. runGuideTabView의 UI로직을 분리하는 편이 좋을 것 같습니다. 현재 선언하신 방식은 다소 가독성이 떨어져 보입니다. 
5. cancelTimerConnect과 didDismiss은 굳이 메서드로 정의하지 않으셔도 될 것 같습니다. 

### 답변

2. selectedCardItem = runGuideExperienceItemArrs[selectedRunGuideTabItem] 이 코드를 이야기 하시는건가요?
onTapGesutre 이벤트 발생시 JustStartView의 runGuideTabView에서 현재 보여지고 있는 항목을 selectedCardItem: RunGuideItem 변수에 저장합니다. 그래야 fullScreenCover(item: Binding<Item>, ...) 에서 item 인수에 Binding 변수인 selectedCardItem 을 넣을 수 있습니다.

3. 비슷한 속성끼리 extension 코드 영역 내에 있도록 변경했습니다. 그리고 runGuideTabView 의 UI와 로직 부분을 분리했습니다.

4. 리팩토링 완료.

5. cancelTimerConnect 는 같은 구문이 반복되는것을 피하기 위해 사용한것입니다. 사용해서 나쁠게 없어 보입니다.
Dismiss를 함수로 만들어 쓰는건 관례처럼 쓰이고 있는듯 합니다.

[질문]
1. selectedCardItem은 JustStartView 초기화시 파라미터로 받는데, 옵셔널타입으로 정의하시고 Nil로 선언하신 이유가 있으신가요?

### 답변

1. fullScreenCover는 (item: Binding<Item>, ...) 의 인자를 갖습니다.
따라서 item 인자에 $<RunGuideItem> 와 같이 입력해줘야 하는데
기본값으로 nil 외에 다른 값을 넣는경우 RunGuideItem 타입의 더미값을 넣어줘야 합니다.
이런 경우 더미값을 넣는것보다 옵셔널을 사용하여 nil로 초기화하는게 좋을듯합니다.

wh5929 commented:
> 
SwiftUI 의 hidden 메서드 정의 내용입니다.


    /// extension View {
    /// Hides this view unconditionally.
    ///
    /// Hidden views are invisible and can't receive or respond to interactions.
    /// However, they do remain in the view hierarchy and affect layout. Use
    /// this modifier if you want to include a view for layout purposes, but
    /// don't want it to display.
    ///
    ///     HStack {
    ///         Image(systemName: "a.circle.fill")
    ///         Image(systemName: "b.circle.fill")
    ///         Image(systemName: "c.circle.fill")
    ///             .hidden()
    ///         Image(systemName: "d.circle.fill")
    ///     }
    ///
    /// The third circle takes up space, because it's still present, but
    /// SwiftUI doesn't draw it onscreen.
    ///
    /// ![A row of circles with the letters A, B, and D, with a gap where
    ///   the circle with the letter C should be.](SwiftUI-View-hidden-1.png)
    ///
    /// If you want to conditionally include a view in the view hierarchy, use
    /// an `if` statement instead:
    ///
    ///     VStack {
    ///         HStack {
    ///             Image(systemName: "a.circle.fill")
    ///             Image(systemName: "b.circle.fill")
    ///             if !isHidden {
    ///                 Image(systemName: "c.circle.fill")
    ///             }
    ///             Image(systemName: "d.circle.fill")
    ///         }
    ///         Toggle("Hide", isOn: $isHidden)
    ///     }
    ///
    /// Depending on the current value of the `isHidden` state variable in the
    /// example above, controlled by the ``Toggle`` instance, SwiftUI draws
    /// the circle or completely omits it from the layout.
    ///
    /// ![Two side by side groups of items, each composed of a toggle beneath
    ///   a row of circles with letters in them. The toggle on the left
    ///   is off and has four equally spaced circles above it: A, B, C, and D.
    ///   The toggle on the right is on and has three equally spaced circles
    ///   above it: A, B, and D.](SwiftUI-View-hidden-2.png)
    ///
    /// - Returns: A hidden view.
    @inlinable public func hidden() -> some View

}

hidden은 layout이 그대로 유지된채로 interact를 하지 못하도록 막아주며 보이지 않게 합니다.
MapView에 hidden 메서드를 사용만해도 cpu 점유율이 큰폭으로 감소됩니다.

EmptyView와 hidden의 사용상 차이점은 인스턴스 해제/생성 유무 입니다.
실행되고 있던 MapView를 인스턴스 해제할 필요는 없어 보입니다.

인스턴스 해제/생성에 필요한 자원이 큰 Map과 같은 경우에는 더더욱 그래보입니다.

추가로 Apple SwiftUI Concepts Tutorials 에서 뷰를 숨길때 적합한 방법에 대해 소개한 페이지 첨부합니다.
[https://developer.apple.com/tutorials/swiftui-concepts/choosing-the-right-way-to-hide-a-view](https://developer.apple.com/tutorials/swiftui-concepts/choosing-the-right-way-to-hide-a-view)

wh5929 commented:
> **RunningView**
> 
> [의견]
> 
> 1. 위의 의견과 같이 리터럴 값을 상수로 선언방식 수정 필요해보입니다.
> 2. 불린 변수명의 일반적인 컨벤션은 is + ~ 입니다. 그리고 버튼과 같이 사용되는데 변수명에 justStartButton등 모든 정보가 들어가있을 필요는 없을 것 같습니다.
> 3. BtnSelector, MenuBtn 메서드나 계산속성, 연산자로 대체하셔도 될 것 같습니다. 타입으로 분리할 필요는 없어보입니다.

### 답변

2. 바로시작 버튼과 러닝가이드 버튼의 구분이 필요해서 bool 변수명에 해당 정보를 기입했기 때문에 바꿀 필요가 없다고 생각합니다.
 컨벤션을 준수하도록 수정하겠습니다.
3. 이 부분은 계속 이야기가 나오고 있는데 동현님이 타입으로 추출하자고 해서 타입으로 만들었던 부분입니다.
추후 변수로 수정할 필요성이 생기면 그 때 수정하도록 하겠습니다.

wh5929 commented:
> **Show+Extension**
> 
> [의견]
> 
> 1. EmptyModifier와 겹칩니다. 특별한 이유가 없다면 EmptyModifier를 사용하셔도 될 것 같습니다.
> 2. Extension은 Helper에 분류하는게 맞아 보입니다.
> 3. Extension 파일의 네이밍은 대상(ex: Int+Extension)+Extension 입니다.

### 답변

2.

hanjaesueng commented on GatheRunner/Utils/Show+Extension.swift:
```diff
@@ -0,0 +1,29 @@
+
+// MARK: - Show
+
+struct Show: ViewModifier {
+    let isVisible: Bool
+
+    @ViewBuilder
+    func body(content: Content) -> some View {
+        if isVisible {
+            content
+        } else {
+            content.hidden()
+        }
+    }
+}
+
+extension View {
+    func show(isVisible: Bool) -> some View {
+        ModifiedContent(content: self, modifier: Show(isVisible: isVisible))
+    }
+}
```
질문
----------------------------
EmptyModifier 와 같은 기능으로 보입니다. 별도로 Show modifier를 사용하신 이유가 있을까요 ?

wh5929 replied:
링크 참고해주세요.
https://github.com/ReleaseAppUntilDie/GatheRunner/pull/58#issuecomment-1276658080

wh5929 commented:
> **RunningGuideView**
> 
> [의견]
> 
> 1. cardItem은 아이템들인데 아이템으로 명명되어 있어 혼동할 수 있을 것 같습니다. 또한 componet 들을 (cardImage, cardTitle등으로) 분리하는게 좋을 것 같습니다.
> 2. 위와 같이 size 관련 상수들은 enum 네임스페이스에 정의하고, runGuideCardItemArrs은 뷰모델을 이용해서 직접 뷰에 전달하면 될 것 같습니다.
> 

### 답변

1. 제안대로 분리하니 코드가 한결 깔끔해졌네요.
2. Apple에서 Combine 프레임워크쪽 리터럴 상수는 enum 네임스페이스에 정의하는것을 표준으로 정했나보네요. 리터럴 상수 정의시 enum 네임스페이스를 사용해야겠습니다.

> [질문]
> 
> 1. 위와 같이 selectedCardItem은 옵셔널타입 및 nil 할당 이유가 궁금합니다.
> 
> **FullScreenCoverView**
> 
> [의견]
> 
> 1. 파일을 분리하는게 더 적합해보입니다.
> 2. 변수명들이 모든 내용을 포함하고 있어, 가독성이 좋지 않습니다.
> 3. 이전 좌표값을 담는 변수들이 매번 초기화되는데 그렇다면 변수로 갖고있거나, 메서드로 계산할 필요도 없어보입니다.
> 4. onChange에 대한 설명을 주석으로 할 필요는 없어보입니다.
> 5. FullScreenCoverView와 TextViewSetForRunGuideBottomView 네이밍은 반드시 수정될 필요가 있어보입니다.
> 6. TextViewSetForRunGuideBottomView와 CloseButton은 타입으로 분리되고, 각각의 size 정보가 따로 관리될 필요가 없어보입니다. 이게 전체적으로 다른 파일에서도 이런 방식으로 구현하셨는데, 결국엔 FullScreenCoverView와 같은 컨테이너뷰에서 size가 결정될텐데 추후에 유지보수나 화면비율을 맞추는데 어려움이 있을 것 같습니다.
> 7. gesture, onChange등 이벤트 연산자 따로 분리하는게 좋아보입니다.
> 

### 답변

'1'. 분리하니 훨씬 깔끔해졌네요.
'3'. 해당 코드 없이는 정상적으로 스크롤 구현을 하지 못합니다.
'4'. Reviewer가 무슨 코드인지 쉽게 알 수 있도록 적어놓았습니다. 가독성을 해치지도, 주석이 코드 분석을 어렵게 하지도 않습니다. 그리고 이러한 간단한 주석 생성에 대해 기존에 합의가 되어 있던 상태였습니다.
'5'. 저는 직관적인 이름이라고 생각합니다. 어떤 부분이 문제되시나요?
'6'. 어떤 size를 얘기하시는건가요? 상위뷰 코드영역에서 padding modifier를 사용해야 유지보수시 혼선을 줄일 수 있다는 말씀이신건가요?
frame 설정과 관련된 부분은 없습니다. padding 설정이라면 자식뷰에서 설정한대로 상위뷰에 적용됩니다.
'7'. onChanged, onEnded 이벤트 Modifier 분리가 어렵네요.

> [질문]
> 
> 1. CloseButton을 최상단 우측에 배치하시려고 하신것 같은데 다른 UI로직으로는 안될까요?
> 2. 메서드로 분리하는 기준이 제각각 인것 같습니다. 제스쳐 연산자에 있는 로직과 setForCardItemOverlayYAxisPosition 메서드로 따로 분리된 로직은 어떤 차이가 있기에, 메서드로 분리되거나 분리하지 않은건가요?
> 3. getForRunGuideBottomDescriptionView은 왜 EmptyView를 리턴하는건가요?

### 답변

1. 더 깔끔하게 구현했습니다. 확인바랍니다.
2. 일관성이 없네요. 분리했습니다.
3. GeometryReader 내부 코드 영역에는 View를 반환해야 하는 객체들이 와야 하는데, 단순하게 좌표값을 구하는 함수(getForRunGuideBottomDescriptionView)에서 반환하는 View가 없으니 EmptyView()를 반환해 에러를 해결했었습니다.
--> 수정 완료.

wh5929 committed:
### Review 내용 반영
1. enum namespace 이용하여 리터럴 관리
2. 각종 뷰, UI / 로직 분리
3. Show+Extension 이름 변경 -> View+Extension / Show 구조체 이름 변경 -> Hide
4. 너무 긴 프로퍼티 이름 축소

wh5929 committed:
### 메서드명 수정
show() -> hide()

wh5929 committed:
### 리팩토링
메인 뷰에서 CloseButton() 정렬하기 위해 컨테이너뷰를 사용했으나, 더 좋은 방법을 찾게 되어 수정.

wh5929 committed:
### 리팩토링
디렉터리 생성 및 파일 정리

DalioKim commented:
> * Reviewer가 무슨 코드인지 쉽게 알 수 있도록 적어놓았습니다. 가독성을 해치지도, 주석이 코드 분석을 어렵게 하지도 않습니다. 그리고 이러한 간단한 주석 생성에 대해 기존에 합의가 되어 있던 상태였습니다.
> * 저는 직관적인 이름이라고 생각합니다. 어떤 부분이 문제되시나요?

우선 반복적으로 이야기가 나오는 이야기인데,  가독성에 대한 보편적인 기준이 있습니다. 기준을 반드시 따르거나 기준이 맞다는건 아니지만 어떤 기준들이 있고, 왜 그런지는 알아야하죠. 

onChange 로직자체가 주석이 필요할정도로 복잡한 로직이 아닙니다. 의미없는 주석으로 꼽히는 것중에 하나가 모든 코드에 주석을 다는건데 onChange에서 스크롤 좌표값을 계산하는건 코드를 통해서 충분히 알 수 있습니다.  이와 같은 이유로 주석을 단다면 코드의 모든 부분에 주석을 달아야하고 당연히 코드를 이해하는데 방해가 되죠. 

저희가 책을 읽을때도 생소하거나 전문적인 정보를 주석으로 따로 저자가 알려주지 그외의 모든 단어에 대해서 주석으로 설명하지 않습니다. 
주석에 관련된 레퍼런스입니다. 
참고 부탁드리겠습니다.

https://nesoy.github.io/articles/2018-01/CleanCode-Comment
https://www.youtube.com/watch?v=IgvNCnb2MzM

변수명에 대한 의견입니다.
사용하신 변수명들은 다음과 같습니다. 
viewPositionYAxisOffset
previousYAxisPositionOfView

1. 해당뷰의 y좌표에 대한 네이밍인데 하나는 viewPositionYAxis이고, 또 다른 하나는 YAxisPositionOfView 이죠. 패턴이 제각각으로 보입니다.
2. 해당뷰와 다른뷰의 Y좌표를 구분하기 위해 각 y좌표변수에 모든 뷰의 네이밍을 명시하는 것도 이상하고, 뷰의 네이밍조차 이상합니다. OfView 또는 view를 보고 FullScreenCoverView의 Y좌표라고 알 수 있을지 의문입니다. 심지어 RunGuideBottomDescriptionView라는 타입의 뷰는 존재하지도 않습니다. 

FullScreenCoverView

fullScreenCover 연산자에 의해 띄어지는 화면입니다. UX기능상의 관점으로 네이밍을 지으신것 같은데, fullScreenCover와 중복된다고 생각됩니다. 그리고 FullScreenCoverView를 통해서 각 cardItem의 상세정보를 보여주는 화면이라는걸 전혀 유추할 수 없습니다.\

보편화된 네이밍 컨벤션에 따르면 중복적인 내용을 담는걸 지양하고 있습니다. 예를들어 album과 albumName 처럼요. 이번에 사용하신 네이밍들이 대부분 이러한 형식으로 명명되지 않았나 생각들어 말씀드렸습니다.

DalioKim commented:
> * 어떤 size를 얘기하시는건가요? 상위뷰 코드영역에서 padding modifier를 사용해야 유지보수시 혼선을 줄일 수 있다는 말씀이신건가요?
>   frame 설정과 관련된 부분은 없습니다. padding 설정이라면 자식뷰에서 설정한대로 상위뷰에 적용됩니다.

이번에 구현하신 모든 부분에서 이와 같은 방식으로 구현되긴했는데 우선 FullScreenCoverView에 한해서만 말씀드리면 

TextViewSetForRunGuideBottomView, CloseButton 모두 타입으로 따로 구분되고  fontSize, padding등의 정보를 따로 정의하셨는데 우선 이 두 타입이 타입으로 구분되어야할 이유를 모르겠습니다. 단순한 버튼과 하단뷰이고 따로 기능도 없고, 재사용될만한 여지도 없어보입니다. 
제 생각에는 private var CloseButton: some View로 분리해도 충분하지 않을까 생각합니다.

그런데 이 2개의 서브뷰를 따로 타입으로 구분하시고, 각각의 UI수치에 관한정보를 각자 관리하고 있습니다. 이 2개의 서브뷰는 FullScreenCoverView에서 쓰이고 FullScreenCoverView에서 최종적인 UI배치가 이뤄지는데 이런 UI 크기에 관련된 정보가 각 서브뷰들이 관리되면 FullScreenCoverView가 UI 수정이 되는 경우, 각각의 타입들도 다시 수정이 이루어져야 됩니다. 

이런 측면에서 1) 타입으로 분리될 필요가 없어보인다. 2)UI 크기 수치와 관련된 정보는 컨테이너 역할인 FullScreenCoverView에서 관리하는게 맞아 보입니다.

DalioKim commented:
> 6\. onChanged, onEnded 이벤트 Modifier 분리가 어렵네요.
> 
> >

네 나중에 진행하셔도 될 것 같습니다.

DalioKim commented:
> SwiftUI 의 hidden 메서드 정의 내용입니다.

네 그러면 show가 아닌 isHidden과 같은 네이밍으로 바꿔주시고, isVisible 라벨이 없어도 바인딩된다는걸 충분히 알 수 있으니 파라미터 라벨에 _을 설정 하면 더 좋을 것 같습니다.

wh5929 commented:
> > SwiftUI 의 hidden 메서드 정의 내용입니다.
> 
> 네 그러면 show가 아닌 isHidden과 같은 네이밍으로 바꿔주시고, isVisible 라벨이 없어도 바인딩된다는걸 충분히 알 수 있으니 파라미터 라벨에 _을 설정 하면 더 좋을 것 같습니다.

동사 hide() 로 정하는 것이 좋아 보입니다. _ 설정 완료.

wh5929 commented:
> > * Reviewer가 무슨 코드인지 쉽게 알 수 있도록 적어놓았습니다. 가독성을 해치지도, 주석이 코드 분석을 어렵게 하지도 않습니다. 그리고 이러한 간단한 주석 생성에 대해 기존에 합의가 되어 있던 상태였습니다.
> > * 저는 직관적인 이름이라고 생각합니다. 어떤 부분이 문제되시나요?
> 
> 우선 반복적으로 이야기가 나오는 이야기인데, 가독성에 대한 보편적인 기준이 있습니다. 기준을 반드시 따르거나 기준이 맞다는건 아니지만 어떤 기준들이 있고, 왜 그런지는 알아야하죠.
> 
> onChange 로직자체가 주석이 필요할정도로 복잡한 로직이 아닙니다. 의미없는 주석으로 꼽히는 것중에 하나가 모든 코드에 주석을 다는건데 onChange에서 스크롤 좌표값을 계산하는건 코드를 통해서 충분히 알 수 있습니다. 이와 같은 이유로 주석을 단다면 코드의 모든 부분에 주석을 달아야하고 당연히 코드를 이해하는데 방해가 되죠.
> 
> 저희가 책을 읽을때도 생소하거나 전문적인 정보를 주석으로 따로 저자가 알려주지 그외의 모든 단어에 대해서 주석으로 설명하지 않습니다. 주석에 관련된 레퍼런스입니다. 참고 부탁드리겠습니다.
> 
> https://nesoy.github.io/articles/2018-01/CleanCode-Comment https://www.youtube.com/watch?v=IgvNCnb2MzM
> 
> 변수명에 대한 의견입니다. 사용하신 변수명들은 다음과 같습니다. viewPositionYAxisOffset previousYAxisPositionOfView
> 
> 1. 해당뷰의 y좌표에 대한 네이밍인데 하나는 viewPositionYAxis이고, 또 다른 하나는 YAxisPositionOfView 이죠. 패턴이 제각각으로 보입니다.
> 2. 해당뷰와 다른뷰의 Y좌표를 구분하기 위해 각 y좌표변수에 모든 뷰의 네이밍을 명시하는 것도 이상하고, 뷰의 네이밍조차 이상합니다. OfView 또는 view를 보고 FullScreenCoverView의 Y좌표라고 알 수 있을지 의문입니다. 심지어 RunGuideBottomDescriptionView라는 타입의 뷰는 존재하지도 않습니다.
> 
> FullScreenCoverView
> 
> fullScreenCover 연산자에 의해 띄어지는 화면입니다. UX기능상의 관점으로 네이밍을 지으신것 같은데, fullScreenCover와 중복된다고 생각됩니다. 그리고 FullScreenCoverView를 통해서 각 cardItem의 상세정보를 보여주는 화면이라는걸 전혀 유추할 수 없습니다.\
> 
> 보편화된 네이밍 컨벤션에 따르면 중복적인 내용을 담는걸 지양하고 있습니다. 예를들어 album과 albumName 처럼요. 이번에 사용하신 네이밍들이 대부분 이러한 형식으로 명명되지 않았나 생각들어 말씀드렸습니다.

### 답변

레퍼런스 확인했습니다.
문제가 될만한 함수나 변수들의 이름을 바꿨습니다. 가능한 한 패턴을 통일하고 가독성에 신경을 썼습니다.

wh5929 commented:
> > * 어떤 size를 얘기하시는건가요? 상위뷰 코드영역에서 padding modifier를 사용해야 유지보수시 혼선을 줄일 수 있다는 말씀이신건가요?
> >   frame 설정과 관련된 부분은 없습니다. padding 설정이라면 자식뷰에서 설정한대로 상위뷰에 적용됩니다.
> 
> 이번에 구현하신 모든 부분에서 이와 같은 방식으로 구현되긴했는데 우선 FullScreenCoverView에 한해서만 말씀드리면
> 
> TextViewSetForRunGuideBottomView, CloseButton 모두 타입으로 따로 구분되고 fontSize, padding등의 정보를 따로 정의하셨는데 우선 이 두 타입이 타입으로 구분되어야할 이유를 모르겠습니다. 단순한 버튼과 하단뷰이고 따로 기능도 없고, 재사용될만한 여지도 없어보입니다. 제 생각에는 private var CloseButton: some View로 분리해도 충분하지 않을까 생각합니다.
> 
> 그런데 이 2개의 서브뷰를 따로 타입으로 구분하시고, 각각의 UI수치에 관한정보를 각자 관리하고 있습니다. 이 2개의 서브뷰는 FullScreenCoverView에서 쓰이고 FullScreenCoverView에서 최종적인 UI배치가 이뤄지는데 이런 UI 크기에 관련된 정보가 각 서브뷰들이 관리되면 FullScreenCoverView가 UI 수정이 되는 경우, 각각의 타입들도 다시 수정이 이루어져야 됩니다.
> 
> 이런 측면에서 1) 타입으로 분리될 필요가 없어보인다. 2)UI 크기 수치와 관련된 정보는 컨테이너 역할인 FullScreenCoverView에서 관리하는게 맞아 보입니다.

해당 내용 수정했습니다.

wh5929 commented:
>  * 어떤 size를 얘기하시는건가요? 상위뷰 코드영역에서 padding modifier를 사용해야 유지보수시 혼선을 줄일 수 있다는 말씀이신건가요?
>    frame 설정과 관련된 부분은 없습니다. padding 설정이라면 자식뷰에서 설정한대로 상위뷰에 적용됩니다.
>  
>  
>  이번에 구현하신 모든 부분에서 이와 같은 방식으로 구현되긴했는데 우선 FullScreenCoverView에 한해서만 말씀드리면
>  TextViewSetForRunGuideBottomView, CloseButton 모두 타입으로 따로 구분되고 fontSize, padding등의 정보를 따로 정의하셨는데 우선 이 두 타입이 타입으로 구분되어야할 이유를 모르겠습니다. 단순한 버튼과 하단뷰이고 따로 기능도 없고, 재사용될만한 여지도 없어보입니다. 제 생각에는 private var CloseButton: some View로 분리해도 충분하지 않을까 생각합니다.
>  그런데 이 2개의 서브뷰를 따로 타입으로 구분하시고, 각각의 UI수치에 관한정보를 각자 관리하고 있습니다. 이 2개의 서브뷰는 FullScreenCoverView에서 쓰이고 FullScreenCoverView에서 최종적인 UI배치가 이뤄지는데 이런 UI 크기에 관련된 정보가 각 서브뷰들이 관리되면 FullScreenCoverView가 UI 수정이 되는 경우, 각각의 타입들도 다시 수정이 이루어져야 됩니다.
>  이런 측면에서 1) 타입으로 분리될 필요가 없어보인다. 2)UI 크기 수치와 관련된 정보는 컨테이너 역할인 FullScreenCoverView에서 관리하는게 맞아 보입니다.

### 답변

확인했습니다.
해당 내용 수정했습니다.
